crates/typst-pdf/src/lib.rs

```diff
@@ -456,6 +459,88 @@ where
     }
 }
 
+struct Emoji {
+    gid: u16,
+    width: f64,
+    image: Frame,
+}
+
+struct EmojiFont {
+    refs: Vec<Ref>,
+    // glyph id, instruction ids
+    // index % 256 is the index in the type3 font
+    emojis: Vec<Emoji>,
+    bbox: Rect,
+}
+
+struct EmojiFontMap {
+    map: HashMap<Font, EmojiFont>,
+    all_refs: Vec<Ref>,
+}
+
+impl EmojiFontMap {
+    fn new() -> Self {
+        Self { map: HashMap::new(), all_refs: Vec::new() }
+    }
+
+    fn items(&self) -> impl IntoIterator<Item = &EmojiFont> {
+        self.map.values()
+    }
+
+    fn take(&mut self) -> Self {
+        let map = std::mem::take(&mut self.map);
+        EmojiFontMap { map, all_refs: self.all_refs.clone() }
```

Comment:
could also `std::mem::take(&mut self.all_refs)`

Reply:
Actually, I can't, because `ctx.emoji_font_map.all_refs` is used at the end of the PDF generation to add Type3 fonts to the list of document fonts. I consider that this clone should be relatively cheap, as the `Vec` should only hold a few `Ref` at most. I am not entirely satisfied by this however, I feel like it is a bit fragile and that you have to understand the whole PDF export logic to know why in one case we `std::mem::take` and in the other we `.clone()`.